src/surge-xt/gui/widgets/LFOAndStepDisplay.cpp

```diff
@@ -414,12 +417,112 @@ void LFOAndStepDisplay::paint(juce::Graphics &g)
     }
     else
     {
-        paintWaveform(g);
+
+        if (!paramsHasChanged())
+        {
+            g.drawImage(*backingImage, getLocalBounds().toFloat(),
+                        juce::RectanglePlacement::fillDestination);
+            paintTypeSelector(g);
+            return;
+        }
+        else
+        {
+            juce::Colour color =
+                juce::Colour((unsigned char)0, (unsigned char)0, (unsigned char)0, 0.f);
+
+            backingImage->clear(backingImage->getBounds());
+        }
+
+        juce::Graphics gr = juce::Graphics(*backingImage);
+        float zoomFloat = (float)zoomFactor / 100.f;
+
+        gr.addTransform(juce::AffineTransform().scale(zoomFloat * 2.f));
+
+        paintWaveform(gr);
+
+        g.drawImage(*backingImage, getLocalBounds().toFloat(),
+                    juce::RectanglePlacement::fillDestination);
+        waveformIsUpdated = false;
     }
 
     paintTypeSelector(g);
 }
 
+bool LFOAndStepDisplay::paramsHasChanged()
+{
+    bool hasChanged = false;
+
+    if (paramsFromLastDrawCall[lfodata->delay.param_id_in_scene].i != lfodata->delay.val.i)
```

Comment:
Happy to merge this, but the lfo data has params which are contiguous so you could also make this code

```
   auto *p = &lfodata->delay; // look in the definition of LFOStorage for which param is first
  while (p <= &lfodata->shape) // and last
   {
      if (paramsFromLastDrawCa[[p->paramid].i != p->val.i;
         hasChanged = true;
   }
   ```
   
   type thing. If you want to make this change now I can wait but I'm also happy to merge.

Thanks for doing this!

Reply:
That is much cleaner.  I was looking for a way of doing that but Couldn't figure out how.  I'll fix it

Reply:
![image](https://github.com/user-attachments/assets/2f6cad1b-fe88-4a5e-b587-c6a037a55819)

Not sure how to interpret the order here. Does this mean that rate is first and release last? Or is there some other list that sets the order of these?

Reply:
rate is first and release is last yes.